File: 14-TypeScript/Grey Multi-Tenant/packages/grey-ruby/lib/grey_sdk/config/options.rb
# frozen_string_literal: true

module GreySdk
  module Config
    # Configuration options for the Grey SDK client
    class Options
      DEFAULT_TIMEOUT = 30
      DEFAULT_PORT = 443

      attr_reader :host, :port, :use_tls, :timeout, :headers

      # @param host [String] Server hostname
      # @param port [Integer] Server port
      # @param use_tls [Boolean] Whether to use HTTPS
      # @param timeout [Integer] Request timeout in seconds
      # @param headers [Hash] Additional headers
      def initialize(host:, port: DEFAULT_PORT, use_tls: true, timeout: DEFAULT_TIMEOUT, headers: {})
        @host = host
        @port = port
        @use_tls = use_tls
        @timeout = timeout
        @headers = headers
      end

      # Creates options for local development
      # @param port [Integer] Local server port
      # @return [Options]
      def self.local(port = 8080)
        new(
          host: "localhost",
          port: port,
          use_tls: false,
          timeout: DEFAULT_TIMEOUT
        )
      end

      # Creates options for production with TLS
      # @param host [String] Production server hostname
      # @param port [Integer] Server port
      # @return [Options]
      def self.production(host, port = 443)
        new(
          host: host,
          port: port,
          use_tls: true,
          timeout: DEFAULT_TIMEOUT
        )
      end

      # Returns the base URL for HTTP requests
      # @return [String]
      def base_url
        scheme = @use_tls ? "https" : "http"
        "#{scheme}://#{@host}:#{@port}"
      end

      # Returns new options with a different timeout
      # @param timeout [Integer] New timeout in seconds
      # @return [Options]
      def with_timeout(timeout)
        Options.new(
          host: @host,
          port: @port,
          use_tls: @use_tls,
          timeout: timeout,
          headers: @headers
        )
      end

      # Returns new options with additional headers merged
      # @param new_headers [Hash] Headers to merge
      # @return [Options]
      def with_headers(new_headers)
        Options.new(
          host: @host,
          port: @port,
          use_tls: @use_tls,
          timeout: @timeout,
          headers: @headers.merge(new_headers)
        )
      end

      # Converts options to a hash
      # @return [Hash]
      def to_h
        {
          host: @host,
          port: @port,
          use_tls: @use_tls,
          timeout: @timeout,
          headers: @headers
        }
      end
    end
  end
end

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-r/R/options.R
#' @title Grey Options
#' @description Configuration options for the Grey SDK
#' @name GreyOptions
NULL

#' @importFrom R6 R6Class
#' @export
GreyOptions <- R6::R6Class(
  "GreyOptions",
  
  public = list(
    #' @field host Server hostname
    host = NULL,
    
    #' @field port Server port
    port = NULL,
    
    #' @field use_tls Whether to use HTTPS
    use_tls = NULL,
    
    #' @field timeout Request timeout in seconds
    timeout = NULL,
    
    #' @field headers Additional headers
    headers = NULL,
    
    #' @description Create new options
    #' @param host Server hostname
    #' @param port Server port (default: 443)
    #' @param use_tls Use HTTPS (default: TRUE)
    #' @param timeout Timeout in seconds (default: 30)
    #' @param headers Additional headers (default: list())
    initialize = function(host, port = 443, use_tls = TRUE, timeout = 30, headers = list()) {
      self$host <- host
      self$port <- port
      self$use_tls <- use_tls
      self$timeout <- timeout
      self$headers <- headers
    },
    
    #' @description Get the base URL
    #' @return Base URL string
    base_url = function() {
      scheme <- if (self$use_tls) "https" else "http"
      paste0(scheme, "://", self$host, ":", self$port)
    },
    
    #' @description Create new options with different timeout
    #' @param timeout New timeout value
    #' @return New GreyOptions
    with_timeout = function(timeout) {
      GreyOptions$new(
        host = self$host,
        port = self$port,
        use_tls = self$use_tls,
        timeout = timeout,
        headers = self$headers
      )
    },
    
    #' @description Create new options with additional headers
    #' @param new_headers Headers to merge
    #' @return New GreyOptions
    with_headers = function(new_headers) {
      merged <- c(self$headers, new_headers)
      GreyOptions$new(
        host = self$host,
        port = self$port,
        use_tls = self$use_tls,
        timeout = self$timeout,
        headers = merged
      )
    }
  )
)

#' Create options for local development
#'
#' @param port Local server port (default: 8080)
#' @return GreyOptions
#' @export
GreyOptions$local <- function(port = 8080) {
  GreyOptions$new(
    host = "localhost",
    port = port,
    use_tls = FALSE,
    timeout = 30
  )
}

#' Create options for production
#'
#' @param host Production server hostname
#' @param port Server port (default: 443)
#' @return GreyOptions
#' @export
GreyOptions$production <- function(host, port = 443) {
  GreyOptions$new(
    host = host,
    port = port,
    use_tls = TRUE,
    timeout = 30
  )
}
